internal/ingress/controller/parser/parser_test.go

```diff
@@ -135,6 +135,488 @@ func TestGlobalPlugin(t *testing.T) {
+		assert.Equal(kong.Configuration{"hide_credentials": true}, state.Plugins[2].Config)
+		assert.Equal("basic-auth", *state.Plugins[3].Name)
+		assert.Equal(kong.Configuration{"hide_credentials": true}, state.Plugins[3].Config)
+	})
+
+	t.Run("plugins with missing secrets or keys are not constructed", func(t *testing.T) {
+		objects := stock
+		objects.KongPlugins = []*configurationv1.KongPlugin{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "global-foo-plugin",
+					Namespace: "default",
+					Labels: map[string]string{
+						"global": "true",
+					},
+				},
+				PluginName: "jwt",
+				ConfigFrom: configurationv1.SecretValueFromSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						Key: "missing-key",
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: "conf-secret",
+						},
+					},
+				},
+			},
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo-plugin",
+					Namespace: "default",
+				},
+				PluginName: "jwt",
+				ConfigFrom: configurationv1.SecretValueFromSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						Key: "missing-key",
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: "conf-secret",
+						},
+					},
+				},
+			},
+		}
+		objects.KongClusterPlugins = []*configurationv1.KongClusterPlugin{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "global-bar-plugin",
+					Labels: map[string]string{
+						"global": "true",
+					},
+				},
+				Protocols:  []string{"http"},
+				PluginName: "basic-auth",
+				ConfigFrom: configurationv1.NamespacedSecretValueFromSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						Key: "basic-auth-config",
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: "missing-secret",
+						},
+					},
+					Namespace: "default",
+				},
+			},
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "bar-plugin",
+				},
+				Protocols:  []string{"http"},
+				PluginName: "basic-auth",
+				ConfigFrom: configurationv1.NamespacedSecretValueFromSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						Key: "basic-auth-config",
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: "missing-secret",
+						},
+					},
+					Namespace: "default",
+				},
+			},
+		}
+		objects.Secrets = []*corev1.Secret{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					UID:       types.UID("7428fb98-180b-4702-a91f-61351a33c6e4"),
+					Name:      "conf-secret",
+					Namespace: "default",
+				},
+				Data: map[string][]byte{
+					"jwt-config":        []byte(jwtPluginConfig),
+					"basic-auth-config": []byte(basicAuthPluginConfig),
+				},
+			},
+		}
+		store, err := store.NewFakeStore(objects)
+		assert.Nil(err)
+		parser := New(store)
+		state, err := parser.Build()
+		assert.Nil(err)
+		assert.NotNil(state)
+		assert.Equal(0, len(state.Plugins),
+			"expected no plugins to be rendered")
+	})
+
+	t.Run("plugins with both config and configFrom are not constructed", func(t *testing.T) {
```

Comment:
Not against this behavior, but what is the rationale here? What if we define precedence instead to ease migration?

Reply:
As designed, the two cannot really coexist: you either use `config` or `configFrom`. Precedence still allows you to use only one or the other, but which you actually use is decided by internal precedence rules. Since they cannot coexist, I think it's good to enforce good config hygiene here and force users to remove whichever they aren't using.

The closest analog I can think of here is the KongIngress/annotation split, which does define precedence, but I think it makes more sense there. KongIngress settings can coexist with annotation settings, since those settings are loaded piecemeal, and you can, for example, set `preserve_host` via an annotation while setting `protocols` in a KongIngress. No such summed config exists with KongPlugin.configFrom and KongPlugin.config; they're all or nothing.

While migration should be a relatively simple (copy-pasta existing config blocks into a secret and write a configFrom), I think it's fine to do something similar via a new KongPlugin if you want to have the original available in case something goes wrong and. Swapping between the new and old is just a matter of adjusting annotation on the other resources.